### PR TITLE
feat(tofu): Inspect globals passed as function arguments

### PR DESCRIPTION
**Pull request by FrederikBolding**
Early attempt at inspecting globals passed as function arguments to effectively narrow global policy. 

Ideal case being able to narrow something like the following and bailing on everything else:
```js
function getHref(doc) {  
  return doc.location.href;
}

const href = getHref(document);

console.log(href);
```

### CODE REVIEW

File: packages/tofu/test/inspectGlobals.spec.js [hunk 1]
```diff
@@ -443,8 +443,208 @@ testInspect(
...
+  },
+  {
+    'console.log': 'read',
+    document: 'read',
+    require: 'read',
+  }
+)
+
+testInspect.bind(test.failing)(
+  'passing global to member-expression callee',
```
Comment (naugtur) [inline packages/tofu/test/inspectGlobals.spec.js:608]: this would be additional 20 lines to support and really hard to know we're not missing an edge case

File: packages/tofu/test/inspectGlobals.spec.js [hunk 2]
```diff
@@ -443,8 +443,208 @@ testInspect(
...
+    console.log(href)
+  },
+  {
+    'console.log': 'read',
+    'document.location.href': 'read',
+  }
+)
+
+testInspect(
+  'naive behavior: attributes usage after parameter reassignment to global argument',
```
Comment (naugtur) [inline packages/tofu/test/inspectGlobals.spec.js:626]: This might be annoying, but would be expensive to handle and maintain IMHO

**Comment by naugtur** [comment]
@FrederikBolding I've added some test that surfaced edge cases and added support for more kinds of function declarations.

File: packages/tofu/src/inspectSource.js [hunk 3]
```diff
@@ -121,8 +134,7 @@ function inspectGlobals(
 
   /**
    * @param {string} variableName
-   * @param {import('@babel/types').Identifier
-   *   | import('@babel/types').ThisExpression} identifierNode
+   * @param {import('@babel/types').Identifier | import('@babel/types').ThisExpression} identifierNode
```
Comment (boneskull) [inline packages/tofu/src/inspectSource.js:137]: (nit)

Better to have

```js
/**
 * @import {Identifier, 
 *   ThisExpression, 
 *   FunctionDeclaration, 
 *   FunctionExpression, 
 *   ArrowFunctionExpression} from '@babel/types'
 */
```

Up at the top instead of repeating `import('@babel/types')...` everywhere. I realize the code doesn't _currently_ use the `@import` pattern, but I'd rather not make it worse. 😄

File: packages/tofu/src/inspectSource.js [hunk 4]
```diff
@@ -158,6 +170,68 @@ function inspectGlobals(
...
+   *   import('@babel/types').Identifier | import('@babel/types').ThisExpression
+   * >} path
+   * @param {import('@babel/types').CallExpression} callExpression
+   * @returns {{ keys: string[]; use: GlobalPolicyValue }[] | null}
+   */
+  function inspectFunctionArgumentUsage(path, callExpression) {
+    const argIndex = callExpression.arguments.indexOf(path.node)
+    if (argIndex === -1 || callExpression.callee.type !== 'Identifier') {
+      return null
+    }
```
Comment (boneskull) [inline packages/tofu/src/inspectSource.js:186]: (nit)

I think this is a little easier to understand

```suggestion
    if (callExpression.callee.type !== 'Identifier') {
       return null
    }
    const argIndex = callExpression.arguments.indexOf(path.node)
    if (argIndex === -1) {
      return null
    }
```

File: packages/tofu/src/inspectSource.js [hunk 5]
```diff
@@ -158,6 +170,68 @@ function inspectGlobals(
...
+    /**
+     * @type {import('@babel/traverse').NodePath<
+     *       | import('@babel/types').FunctionDeclaration
+     *       | import('@babel/types').FunctionExpression
+     *       | import('@babel/types').ArrowFunctionExpression
+     *     >
+     *   | null}
+     */
+    let functionPath = null
+    if (bindingPath?.isFunctionDeclaration()) {
```
Comment (boneskull) [inline packages/tofu/src/inspectSource.js:197]: (nit)

seems like we could short-circuit some logic here by returning `null` if `bindingPath` is nullish

File: packages/tofu/src/inspectSource.js [hunk 6]
```diff
@@ -158,6 +170,68 @@ function inspectGlobals(
...
+
+    const param = functionPath.node.params[argIndex]
+    // skip parameters that arent plain identifiers
+    if (param?.type !== 'Identifier') {
+      return null
+    }
+    const paramReferences = functionPath.scope.getBinding(
+      param.name
+    )?.referencePaths
+    if (!paramReferences || paramReferences.length === 0) {
```
Comment (boneskull) [inline packages/tofu/src/inspectSource.js:220]: (nit/style)

```suggestion
    if (!paramReferences?.length) {
```

File: packages/tofu/src/inspectSource.js [hunk 7]
```diff
@@ -158,6 +170,68 @@ function inspectGlobals(
...
+      return null
+    }
+    return paramReferences.map((ref) => {
+      const { path, identifierUse } = inspectIdentifierForDirectMembershipChain(
+        param.name,
+        /** @type {import('@babel/types').Identifier} */ (ref.node),
+        /** @type {import('./inspectPrimordialAssignments').MemberLikeExpression[]} */ (
```
Comment (boneskull) [inline packages/tofu/src/inspectSource.js:227]: Q: Why are these type assertions needed?

File: packages/tofu/test/inspectGlobals.spec.js [hunk 8]
```diff
@@ -443,8 +443,208 @@ testInspect(
...
+  {
+    'console.log': 'read',
+    'document.location.href': 'read',
+  }
+)
+
 function testInspect(label, opts, fn, expectedResultObj) {
-  test(label, (t) => {
+  const run = typeof this === 'function' ? this : test
+  run(label, (t) => {
```
Comment (boneskull) [inline packages/tofu/test/inspectGlobals.spec.js:647]: please add another parameter (e.g., `testFn: TestFn`) instead of depending on the context (which is not surfaced anywhere).  tbh this should be refactored into a macro anyway

File: packages/tofu/src/inspectSource.js [hunk 9]
```diff
@@ -158,6 +170,68 @@ function inspectGlobals(
...
+  /**
+   * Collects keys used from a global in a function it is being passed to
+   *
+   * @param {import('@babel/traverse').NodePath<
+   *   import('@babel/types').Identifier | import('@babel/types').ThisExpression
+   * >} path
+   * @param {import('@babel/types').CallExpression} callExpression
+   * @returns {{ keys: string[]; use: GlobalPolicyValue }[] | null}
+   */
+  function inspectFunctionArgumentUsage(path, callExpression) {
```
Comment (boneskull) [inline packages/tofu/src/inspectSource.js:182]: (could/style)

I'd prefer the API always returned an array, which should eliminate nullish checks elsewhere.

**Review by boneskull — COMMENTED** [review]
Thanks!!

I don't have any hills to die on here, but would really prefer:

- Change to the `testInspect` signature
- Change to the `inspectFunctionArgumentUsage` signature
